### ISSUE TITLE
[RFC]: Inter-Request DiT Cache with Semantic Reuse

### ISSUE THREAD

**Issue by cta-11** [vllm-project/vllm-omni#3978]
### Motivation.

**1.1 Background** 
The vLLM-Omni framework currently provides two caching acceleration mechanisms, but both are intra-request optimizations:
- CacheDiT (cache_dit): Caches the intermediate features of the DiT transformer during a single denoising process, skipping redundant computations. 
- TeaCache (tea_cache): Based on time step similarity, skips transformer forward for certain steps during a single denoising process. These solutions share a common limitation: the cache is only valid within the lifecycle of a single request. 
Once the request ends, the cache is discarded, and the next request, even if it has highly similar semantics, must perform a complete denoising process from scratch.

**1.2 Present Situation** 
In practical production environments, users often submit prompts with similar semantics. The denoising process for these requests is highly consistent in the early steps, where the global structure (layout, subject, tone) is determined, with the later steps adjusting details. However, existing frameworks cannot leverage this cross-request similarity, and each time a complete inference is required, resulting in significant waste of GPU computing resources.

**1.3 Solution Approach**
Based on similarity matching of prompts, we retrieve semantically similar prompts, resume from the intermediate denoising step, and skip the early steps that have already converged.

### Proposed Change.

**2.1.1 Exact Match Cache Design**
Hash all deterministic parameters of the request as the cache key:

```
@dataclass
class CacheKey:
    prompt: str
    negative_prompt: str
    height: int
    width: int
    num_inference_steps: int
    guidance_scale: float
    true_cfg_scale: float
    seed: int
    sigmas: tuple[float, ...] | None
    max_sequence_length: int | None
    num_images_per_prompt: int

    def to_hash(self) -> str:
```

The CacheKey is constructed from the request object using build_cache_key_from_request(req, pipeline), automatically resolving None parameters (such as height/width) to the pipeline default values.

**2.1.2 Cache storage (DiTCacheStore)**
An LRU cache based on OrderedDict, supporting:
- Memory limit: max_memory_gb (default 4GB), automatically evicting the oldest unused entries when exceeded
- Entry limit: max_entries (default 100)
- Thread safety: threading.Lock protects all read and write operations
- Hit statistics: Record hits/misses/hit_rate
Each cache entry stores:
- latents: The final latent (the result after denoising, before VAE decoding)
- step_latents: The intermediate latent for each denoising step (used for recovery from intermediate steps)
- cache_key_hash: Cache key hash
- metadata: Optional metadata
All tensors are stored on the CPU using .detach().clone().cpu(), without occupying GPU memory. Each entry is approximately 31MB (1 final + 50 step latents).
However, this is just a prototype with simple implementation. In the future, we will consider integrating DiT Cache storage with existing storage mechanisms (LMCache, mooncake) to store the cache in DRAM and SSD.

**2.1.3 Step Latents Recorder**
Attach to the pipeline (pipeline._step_latents_recorder) and record the latent after each step of the denoising loop.
```
step_latents_recorder = getattr(self, "_step_latents_recorder", None)
for i, t in enumerate(timesteps):
    ...
    latents = self.scheduler_step_maybe_with_cfg(noise_pred, t, latents, do_true_cfg)
    if step_latents_recorder is not None:
        step_latents_recorder.record(i, t.item(), latents)

```
**2.2 Semantic Matching Design** 
Referencing the implementation method of stage1 in the paper "Beyond Few-Step Inference: Accelerating Video Diffusion Transformer Model Serving with Inter-Request Caching Reus", we utilize the CLIP model for semantic matching of text.

**2.2.1 CacheEntry extension**
Add two new fields to CacheEntry:
```
@dataclass
class CacheEntry:
    latents: torch.Tensor
    cache_key_hash: str
    step_latents: list[StepLatentData] | None = None
    metadata: dict[str, Any] | None = None
    clip_embedding: torch.Tensor | None = None
    cache_key: CacheKey | None = None
```
- clip_embedding: The CLIP pooler output (768-dimensional) of the prompt, used for semantic search. 
- cache_key: Retains the original request parameters for parameter compatibility filtering. 
When calling store(), the CLIP embedding is automatically encoded and stored.

**2.2.2 Semantic search (semantic_search)**
Add a new method in DiTCacheStore to traverse all cache entries for cosine similarity matching:
```
def semantic_search(
    self,
    query_embedding: torch.Tensor,
    threshold: float = 0.75,
    target_device: torch.device | str | None = None,
    required_height: int | None = None,
    required_width: int | None = None,
    required_num_inference_steps: int | None = None,
) -> tuple[torch.Tensor | None, list[StepLatentData] | None, float]:
```
Parameter compatibility filtering:
- The height and width must match precisely (latent spaces of different sizes are different and cannot be reused)
- num_inference_steps must be ≥ the number of steps requested by the target (to ensure sufficient intermediate latents in the cache)
All computations are performed on the CPU, with a single search taking approximately 1ms (for 100 entries)

**2.2.3 CLIP encoder**
Load the CLIP text encoder to the CPU in backend.py:
```
def _init_clip_encoder(self) -> None:
    from transformers import CLIPTextModel, CLIPTokenizer
    self._clip_device = torch.device("cpu")
    self._clip_tokenizer = CLIPTokenizer.from_pretrained(self._clip_model_path)
    self._clip_model = CLIPTextModel.from_pretrained(self._clip_model_path)
    self._clip_model.to(self._clip_device)
    self._clip_model.eval()
```
- Only using text encoder (no vision encoder required), ~1.5GB CPU memory 
- single encoding ~1ms, without increasing inference latency

**2.2.4 Dynamic skip mapping**
Calculate the number of skipped steps dynamically based on similarity (as a monotonically increasing function):
_ratio = (similarity - threshold) / (1.0 - threshold)
skip_steps = min_skip + ratio × (max_skip - min_skip)_

**2.3Future work**

**2.3.1Prompt differential analysis**
After completing the full reuse stage, perform single-step cache partial reuse based on the difference degree of prompts (the directional guidance model is enhanced for prompts with changes in key attributes).

**2.3.2 From text-to-image model to text-to-video model**

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by hsliuustc0106** [comment]
cc @ZJY0516 @yuanheng-zhao @wtomin @SamitHuang

**Comment by yuanheng-zhao** [comment]
I think this feature is worth discussing and exploring. From the reference paper, their Chorus caching approach contains 3 stages as a whole:

1. Pure full reuse of matched source latents (CLIP retrieval + skip boundary K1)
2. Selective Region Denoising (partial reuse) + TGAA (Token-Guided Attention Amplification) for semantic realignment
3. Full compute / repair

Reference: 
Beyond Few-Step Inference: Accelerating Video Diffusion Transformer Model Serving with Inter-Request Caching Reuse: https://arxiv.org/pdf/2604.04451

**Comment by yuanheng-zhao** [comment]
Several concerns after reviewing this proposal,

1. about compatibility filter

- Consistency of steps: what if the stored latents are from a request with m steps, and a coming request is with n steps. In these scenarios, the cache seems not to be easily replaced to initial steps for quality retention.

> num_inference_steps must be ≥ the number of steps requested by the target 

- if there're scheduler states to maintain, for different models
- might want to determine other factors to count in

2. memory occupancy: For a non-time-step-distillation model running full denoising steps (e.g., 40, 50 steps), it occupies non-trivial device memory for storing all the latents of each step. For example, latents of Qwen-Image running for a 50-steps and 1328×1328 image cost about 40MB. 4GB memory limit serves for around 100 requests, which seems to have little chance to got hit by any semantically similar requests in production env. 


@cta-11  We could discuss the concerns and APIs, revising the initial implementation plan, and leave some extensions for future TODO

**Comment by cta-11** [comment]
@yuanheng-zhao Thank you for your recognition. Regarding the questions you raised, here are my thoughts and those of my team:

1. If the number of inference steps required by the newly arrived request is greater than that of the matched request (i.e., n > m), the new request will execute the full number of inference steps and save the latent state. If n < m, the inference will start from the m * alpha step, based on the calculated reuse ratio alpha. (The actual effect needs to be verified.)

2. Regarding the memory usage issue, we will refer to the uninstallation mechanism of kvcache and interconnect with the backend (such as LMCache and mooncake) to increase the cache size.

3. The current demo mainly implements stage 1 mentioned in the Chorus paper. In the future, we also hope to implement some reuse functions in stage 2 in vllm-omni.

**Comment by yuanheng-zhao** [comment]
@cta-11 Got it, thanks for explaining. I left several more notes on the proposed changes and your initial design, specifically and for general suggestions: 

1. The placement of `_step_latents_recorder `: We might prefer capture latens in a shared method/mixin/ModelHook rather than inserting and editing each denoising loop. Consider utilizing hook `callback_on_step_end` or creating some shared method / mixin.
2. Ensure a proper and consistent design on cache key and semantic search factors, for example, I noticed there exists seed i cache key, and semantic search might require refined constraints.
3. Let's define the scope that inter-request cache should handle and will not handle. For example, LoRA adapter, quantization, etc - I think these affect the cacheing choices; and continuous batching (stepwise batching) might be orthogonal to this feat.
4. **Distributed, parallelism scenarios**(Maybe future plan): With deploy scenarios like TP/SP/PP, it becomes a bit complicated to manage replicated latents.

**Comment by yuanheng-zhao** [comment]
An example place to make this feature as a self-contained subpackage: `vllm_omni/diffusion/cache/inter_request/`

**Comment by cta-11** [comment]
@yuanheng-zhao Thank you for your suggestion. My team and I need to reorganize the code, and we will submit a code PR in a few days.

**Comment by asukaqaq-s** [comment]
Is there any expected performance gain or accuracy loss? This seems basically useless for distilled models.

**Comment by yx0716** [comment]
Hi @cta-11 — is the PR mentioned above still on the way? We'd love to see this land, and would be glad to work on it together if extra hands are useful.

For context: we've been working on the same problem for our own video-generation serving, and open-sourced our attempt as [CacheSeek](https://github.com/Tele-AI/CacheSeek) — cross-request reuse for video DiT (semantic retrieval + resume from intermediate-step latent checkpoints), currently running with Wan2.2-14B T2V and LTX 2.3. It's still alpha, but in our tests it gave a 23–36% E2E latency cut on Wan2.2, stacked on top of in-request caches (cache-dit / TeaCache).

A few things we ran into that echo the discussion above, in case they're useful data points:

- memory & hit rate: we ended up checkpointing only a few key steps per request and moving latents to tiered storage (local FS / remote) — an in-memory-only corpus was too small to get meaningful hit rates for us;
- +1 to `callback_on_step_end` for recorder placement;
- hits needed a quality gate beyond text-embedding similarity — we added a rerank step, otherwise some hits were visually wrong.

Happy to join forces on the vllm-omni integration — contributing PRs, eval data, or just sharing what we've learned, in whatever split makes sense. cc @yuanheng-zhao

**Comment by cta-11** [comment]
@yx0716 Currently, PR is still ongoing. 
In the recent period, we have mainly been debugging the adaptation of the text-to-image model (based on the qwen-image model). The issues we encountered are similar to what you mentioned: if we only perform similarity matching based on the text, some images appear incorrect. Therefore, while saving the latent state, we also saved the embeddings of previously generated images and performed a similarity match between the current request prompt and the images generated from the matched request prompts. We used this value as a penalty coefficient, which effectively reduced the aforementioned issues. However, this approach introduced another problem: the number of reuse steps for most requests was reduced. This issue can be addressed later using other methods. At the same time, we have also made adaptations for cache-dit and other features, allowing them to be enabled simultaneously. 

Currently, we are conducting code reviews and expect to push the changes to the community in the near future. 

Another focus of ours is to extend the above approach to the text-to-video domain, using the Wan2.2-14B T2V model. However, the penalty coefficient is no longer effective in this case, so we are currently exploring other methods to address the issue of some videos appearing distorted.

**Comment by yx0716** [comment]
Thanks for the update. The image side sounds solid. Since our focus is specifically on video, we'd love to direct our contributions there.

On sequencing: for video, [FlexCache](https://arxiv.org/abs/2501.04012) and [Chorus](https://arxiv.org/abs/2604.04451) are both published but neither ships a public implementation. Given this, we were wondering if it makes sense to expose stage-1 full reuse as an opt-in feature (default off, like `cache_backend`) for now. Our VLM-judge QA on video pairs shows the quality drop is real but bounded, so an opt-in flag could let users choose the **trade-off** while we keep iterating on the quality side.

From our video runs: the earliest steps fix the global layout, so reuse depth K is the sensitive variable — in a forced-K sweep on Wan2.2-14B T2V (~590 pairs, VLM-judged), K=3 tolerated rerank ≈0.63 while K≥7 needed ≈0.85+; we ended up picking K per donor from a small threshold table. Net: **23–36% E2E** at acceptable quality.

Pipeline-wise, we needed per-step latent capture and resume-from-step-k in Wan2.2. Building on existing work, we are also developing an integration for `vllm-omni`. We are currently validating this on a dual-GPU Hopper setup, and we'd love to work closely with the team to upstream this integration, ensuring it fully aligns with `vllm-omni`'s compliance guidelines.
